### PR TITLE
Make sure to update the cache both by id and key

### PR DESCRIPTION
**Pull request by kjac**
### Prerequisites

- [x] I have [created an issue](https://github.com/umbraco/Umbraco-CMS/issues) for the proposed changes in this PR, this fixes: https://github.com/umbraco/Umbraco-CMS/issues/2997
- [x] I have added steps to test this contribution in the description below

### Description

This PR ensures that the entity cache is updated for both entity id and key. 

#2997 outlines the steps to reproduce. The following test template can also be used to test (note that you need to update the id and key of the content item before testing):

```cshtml
@inherits Umbraco.Web.Mvc.UmbracoTemplatePage
@{
	Layout = null;
	// the id and key of the content to test
	var id = 1106;
	var key = Guid.Parse("2886e85f-d8cc-4ea7-9524-b2009490031e");
}
<html>
    <body>
        <strong>Version ID from ContentService:</strong><br/>
        @UmbracoContext.Current.Application.Services.ContentService.GetById(id).Version<br>
        @UmbracoContext.Current.Application.Services.ContentService.GetById(key).Version<br>
    </body>
</html>
```

### CODE REVIEW

**Comment by emmaburstow** [comment]
Hey @kjac 

Thanks for this! We'll have a look asap and let you know if we need a thing. Happy Hacktoberfest!

Em

**Comment by nul800sebastiaan** [comment]
Hey @kjac - this might have some unintended consequences since I can see some unit tests are failing due to this: https://ci.appveyor.com/project/Umbraco/umbraco-cms-b2cri/builds/19515503/tests

Could you look at least at why that might be? I'll ask the team if they have suggestions too.

**Comment by Shazwazza** [comment]
When things get put into the cache, they also need to be removed from the cache with ICacheRefreshers. I'm also pretty sure we are not doing this because this will add 2x the amount of memory to the cache which we really don't want. We have an Id -> Key cache map already for this scenario which is a much smaller memory footprint than adding 2x the amount of objects to the cache. This key mapping is what should be used when looking up items by Key. In many places this is already being done so it might just be missing from some other places. Would be good for @zpqrtbnk  to have a look too.

**Comment by kjac** [comment]
@Shazwazza makes a good point in terms of memory usage, even if the cache TTL is only 5 minutes.

I'm going to try a different approach where the cache is explicitly cleared instead of updated for the entity key. Since the entity id is the primary lookup (at least that's what the code base assumes), this makes sense and should ensure a deterministic behavior by the cache.

**Comment by kjac** [comment]
@Shazwazza makes a good point in terms of memory usage, even if the cache TTL is only 5 minutes.

I'm going to try a different approach where the cache is explicitly cleared instead of updated for the entity key. Since the entity id is the primary lookup (at least that's what the code base assumes), this makes sense and should ensure a deterministic behavior by the cache.

**Comment by Shazwazza** [comment]
When things get put into the cache, they also need to be removed from the cache with ICacheRefreshers. I'm also pretty sure we are not doing this because this will add 2x the amount of memory to the cache which we really don't want. We have an Id -> Key cache map already for this scenario which is a much smaller memory footprint than adding 2x the amount of objects to the cache. This key mapping is what should be used when looking up items by Key. In many places this is already being done so it might just be missing from some other places. Would be good for @zpqrtbnk  to have a look too.

**Comment by kjac** [comment]
@Shazwazza makes a good point in terms of memory usage, even if the cache TTL is only 5 minutes.

I'm going to try a different approach where the cache is explicitly cleared instead of updated for the entity key. Since the entity id is the primary lookup (at least that's what the code base assumes), this makes sense and should ensure a deterministic behavior by the cache.

**Comment by Shazwazza** [comment]
Have a look at the usage of https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Core/Services/IdkMap.cs in the source, it's the int -> guid cache that can be used so that the repo caches are only cached based on their PK but this cache can be used to convert between the 2 keys

**Comment by Shazwazza** [comment]
(GH is not allowing me to comment, sorry if you get this multiple times in email)

Have a look at the usage of https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Core/Services/IdkMap.cs in the source, it's the int -> guid cache that can be used so that the repo caches are only cached based on their PK but this cache can be used to convert between the 2 keys

**Comment by kjac** [comment]
@Shazwazza GH is having a bit of trouble today :) I've got a solution which I'll post as soon as things are running smoothly again.

**Comment by kjac** [comment]
@Shazwazza GH is having a bit of trouble today :) I've got a solution which I'll post as soon as things are running smoothly again.

**Comment by kjac** [comment]
@Shazwazza ok. New approach as explained earlier (many times apparently, due to GH trouble 😁). Instead of adding the extra item to the cache by key, the cache is explicitly cleared by key instead, thus ensuring the cache is at least not invalid.

I've also taken the liberty to refactor the class a bit to remove the duplicate code.

File: src/Umbraco.Core/Cache/DefaultRepositoryCachePolicy.cs [hunk 1]
```diff
@@ -51,6 +51,34 @@ protected virtual void InsertEntity(string cacheKey, TEntity entity)
...
+            if (entity.HasIdentity == false)
+            {
+                return;
+            }
+
+            // insert the item by id (primary lookup)
+            InsertEntity(GetEntityCacheKey(entity.Id), entity);
+
+            // make sure the item is not cached by key (secondary lookup) to avoid old version lingering in the cache (see #2997)
+            Cache.ClearCacheItem(GetEntityCacheKey(entity.Key));
```
Comment (nul800sebastiaan) [inline src/Umbraco.Core/Cache/DefaultRepositoryCachePolicy.cs]: I'm probaby being dumb, but I don't see this code in the old code, why is this now required?

Comment (kjac) [reply src/Umbraco.Core/Cache/DefaultRepositoryCachePolicy.cs]: @nul800sebastiaan you're not being dumb :) I have basically reverted the first code changes and redone the PR. That plus the GH trouble today for sure isn't helping.

The core of the issue is that the cache by key is not updated when the cache by id is updated (e.g. on create or save). The first time around I forcefully updated the cache by key alongside the cache by id. But as @Shazwazza pointed out, that explicitly doubled the memory allocated by the cache. 

This time around I'm forcefully flushing the cache by key whenever the cache by id is updated. That forces a reload of the entity when it's subsequently requested by key, thus solving the issue.

**Comment by nul800sebastiaan** [comment]
Thanks again @kjac - I could've sworn I commented on this already but it must have been a chat we had at DK festival instead. I'm going to close this one for now as I got feedback from @zpqrtbnk that this is not the way to go. 
There's some weird things going on that he would like to investigate to get to the root cause of the problem here. He was also concerned about cacherefreshers not working with this update.

**Pull request closed without merging**